src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -285,21 +526,95 @@ private static bool OperationUsesPreviewFeatures(OperationAnalysisContext contex
+        {
+            foreach (ITypeSymbol typeParameter in typeArguments)
+            {
+                if (SymbolIsAnnotatedAsPreview(typeParameter, requiresPreviewFeaturesSymbols, previewFeatureAttribute))
+                {
+                    previewSymbol = typeParameter;
+                    return true;
+                }
+            }
+
+            previewSymbol = null;
+            return false;
         }
 
-        private static bool ProcessPreviewAttribute(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, INamedTypeSymbol? previewFeatureAttribute)
+        private static bool SymbolIsAnnotatedAsPreview(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, INamedTypeSymbol previewFeatureAttribute)
         {
+            static bool ContainingSymbolHeirarchyIsAnnotatedAsPreview(ISymbol symbol,
+                ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols,
+                INamedTypeSymbol previewFeatureAttribute, [NotNullWhen(true)] out ISymbol? containingPreviewSymbol)
+            {
+                INamedTypeSymbol? containingType = symbol.ContainingType;
+                while (containingType != null)
+                {
+                    // Namespaces do not have attributes
+                    if (!(containingType is INamespaceSymbol))
+                    {
+                        if (containingType.HasAttribute(previewFeatureAttribute))
+                        {
+                            requiresPreviewFeaturesSymbols.GetOrAdd(containingType, true);
+                            containingPreviewSymbol = containingType;
+                            return true;
+                        }
+                    }
+
+                    containingType = containingType.ContainingType;
+                }
+
+                containingPreviewSymbol = null;
+                return false;
+            }
+
             if (!requiresPreviewFeaturesSymbols.TryGetValue(symbol, out bool existing))
             {
-                bool ret = symbol.HasAttribute(previewFeatureAttribute);
-                requiresPreviewFeaturesSymbols.GetOrAdd(symbol, ret);
-                return ret;
+                if (symbol.HasAttribute(previewFeatureAttribute))
+                {
+                    requiresPreviewFeaturesSymbols.GetOrAdd(symbol, true);
+                    return true;
+                }
+
+                if (ContainingSymbolHeirarchyIsAnnotatedAsPreview(symbol, requiresPreviewFeaturesSymbols, previewFeatureAttribute, out ISymbol? containingPreviewSymbol))
+                {
+                    ISymbol loopSymbol = symbol.ContainingSymbol;
+                    requiresPreviewFeaturesSymbols.GetOrAdd(loopSymbol, true);
+                    while (loopSymbol != containingPreviewSymbol)
```

Comment:
> i think you can do it with one loop inside ContainingSymbolHeirarchyIsAnnotatedAsPreview

How though? For ex: let's say we have

```csharp
    [RequiresPreviewFeatures]
    public class NestedClass0
    {
        public class NestedClass1
        {
            public class NestedClass2
            {
                public class NestedClass3
                {
                    public bool AMethod() => false;
                    public bool AProperty => false;
                    public bool AField = true;
                }
            }
        }
    }
```
and we are processing `NestedClass3`. I'm going up the chain in `ContainingSymbolHeirarchyIsAnnotatedAsPreview` until we see the `[RequiresPreviewFeatures]` attribute on `NestedClass0` and return that as the containingPreviewSymbol. We have to traverse the heirarchy again from `NestedClass3` to `NestedClass0` to set the bool flag for all the symbols right? I don't see how we can do it in 1 loop inside `ContainingSymbolHeirarchyIsAnnotatedAsPreview` because we don't know which symbol above the current symbol has a preview attribute?

Reply:
> How though?

Recursively, see how it is done in PCA
https://github.com/dotnet/roslyn-analyzers/blob/51ac1b599127833c7da4622b419c869daab54136/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/PlatformCompatibilityAnalyzer.cs#L1408-L1435

Reply:
Oh the recursion is just using an implied stack, so it's very similar. I fixed mine to use recursion too!

Reply:
Thanks, yes it is similar but i believe recursive will be more efficient